prime-router/src/main/kotlin/CsvConverter.kt

```diff
@@ -8,30 +8,80 @@ import java.io.OutputStream
 /**
  * A converter differs from a serialization in that
  */
-object CsvConverter {
-    private data class Mapping(
+class CsvConverter(val metadata: Metadata) {
+    private data class CsvMapping(
         val useCsv: Map<String, Element.CsvField>,
         val useMapper: Map<String, Pair<Mapper, List<String>>>,
         val useDefault: Map<String, String>,
+        val errors: List<String>,
+        val warnings: List<String>,
     )
 
-    fun read(schema: Schema, input: InputStream, source: Source): Report {
-        return read(schema, input, listOf(source))
+    private data class RowResult(
+        val row: List<String>,
+        val errors: List<String>,
+        val warnings: List<String>,
+    )
+
+    data class ReadResult(
+        /**
+         * The report generated by the read may be empty because of read validation errors
+         */
+        val report: Report?,
+        /**
+         * The list of rows that were rejected while reading
+         */
+        val rejected: List<Int> = emptyList(),
+        /**
+         * The list of errors that caused a item to not be reported
+         */
+        val errors: List<String> = emptyList(),
```

Comment:
I like your idea. Here's what I shared at today's engineering sync for a format. 
```
{
    "id": null,
    "destinations": [],
    "errors": [
        {
            "scope": "REPORT",
            "id": "",
            "details": "Missing 'Result_ID' header"
        },
        {
            "scope": "REPORT",
            "id": "",
            "details": "Missing 'Patient_state' header"
        }
    ],
    "warnings": [
        {
            "scope": "REPORT",
            "id": "",
            "details": "Missing 'Patient_email' header"
        },
        {
            "scope": "REPORT",
            "id": "",
            "details": "Missing 'Ordering_facility_email' header"
        },
        {
            "scope": "ITEM",
            "id": "563423",
            "details": "'Patient_lookup_ID' value is invalid"
        },
        {
            "scope": "REPORT",
            "id": "",
            "details": "Unexpected 'Testing_lab_specimen_ID' header is ignored"
        }
    ]
}
```

Reply:
Yes, this seems ok.   Suggested further improvements for a future PR, and a bug:

- I think people will want to see report level problems first, then row-level.    So, a simple suggestion is to sort the ones with REPORT scope to the top of the `errors` array, and similar for `warnings`.   Keep in mind that the first time someone sees this, its likely to be a gigantic daunting list.

- I get why you are not putting in row numbers, because other formats don't have rows, but rownums sure would be easier on the reader.    At minimum, lets at least make sure all the ITEM level issues appear in the same order as the rows in the file.

- A good readability improvement for the user would be to explicitly use the column name used by the sender for the id.  For example, the primary key for a row in Simple Report is called the Result_ID in the file, so...
```
        {
            "scope": "ITEM",
            "id": "Result_ID:563423",      # Or alternatively simply    "Result_ID" : "563423"
            "details": "'Patient_lookup_ID' value is invalid"
        },
```

- And for extra bonus points and badges:  If _every_ ITEM in a column has the same error, report that as a REPORT level issue, rather than repeating the same error a zillion times.   This will be very common - if the phone number is misformatted, its likely to be globally wrong.


- Oooh, and I just found a bug while looking at this:   pdi-covid-19 has
    `trackingElement: testing_lab_specimen_id`
but `testing_lab_specimen_id` is no longer even in the data!   That should be an error on metadata load.

Reply:
Good stuff. Let's pull this in another PR.